### PR TITLE
Session Hotfixes

### PR DESCRIPTION
**Pull request by dwolfhub**
Hotfixes to SessionProvider->parseConnections and redis session handler.

Setting up Redis for sessions and i had to extend the SessionProvider and the RedisHandler to make hotfixes. Figured I would submit a patch as well.

### CODE REVIEW

**Comment by GawainLynch** [comment]
Test failures are unrelated and fixed in #5557

File: src/Session/Handler/RedisHandler.php [hunk 1]
```diff
@@ -25,7 +25,7 @@ class RedisHandler implements \SessionHandlerInterface, LazyWriteHandlerInterfac
      */
     public function __construct($redis, $maxlifetime)
     {
-        if (!$redis instanceof Redis || !$redis instanceof Predis) {
+        if (!$redis instanceof Redis and !$redis instanceof Predis) {
             throw new \InvalidArgumentException('Argument must be an instance of Redis or Predis\ClientInterface');
```
Comment (GawainLynch) [inline src/Session/Handler/RedisHandler.php]: Style nitpick, but could you flip the `and` to `&&` for code base consistency.

**Comment by GawainLynch** [comment]
@dwolfhub does this achieve what you were needing for #5538 / #5534 ?

**Comment by dwolfhub** [comment]
It's a step in the right direction. You still can't change session options without extending Bolt\Application. I have another PR coming for that.

File: src/Session/Handler/RedisHandler.php (hunk 1 above)
Comment (dwolfhub) [inline src/Session/Handler/RedisHandler.php]: sure thing, made that change

**Comment by CarsonF** [comment]
It's been awhile..what does this change for redis handler?
I'm pretty sure this breaks several options: `persistant`, `database`, `prefix`, etc.

File: src/Provider/SessionServiceProvider.php [hunk 2]
```diff
@@ -341,9 +341,9 @@ protected function parseConnections($options, $defaultHost, $defaultPort)
             }
         } elseif (isset($options['save_path'])) {
             foreach (explode(',', $options['save_path']) as $conn) {
-                $conn = new ParameterBag($conn);
-                $conn->set('uri', new Uri($conn));
-                $toParse[] = $conn;
+                $connBag = new ParameterBag();
+                $connBag->set('uri', new Uri($conn));
+                $toParse[] = $connBag;
             }
```
Comment (dwolfhub) [inline src/Provider/SessionServiceProvider.php]: @CarsonF an explanation: `$conn` is a string, but the constructor for Parameter bag expects an array. Also, you were passing the parameter bag into the `Uri` constructor and then setting that to 'uri' in the same ParameterBag. It didn't make sense to me (and threw an exception), but maybe I'm missing something.

**Comment by dwolfhub** [comment]
@CarsonF it shouldn't, this is only when parsing save_paths for things like redis.

**Comment by CarsonF** [comment]
Yeah ok maybe it was a derp. I think I've seen in php.ini where those other other options will be query string parameters on the url/save_path. So maybe they need to be pulled out after parsing `Uri` and set on `$connBag`?

```
localhost:6379?prefix=bolt_&database=2
```

**Comment by dwolfhub** [comment]
will look into that

File: src/Provider/SessionServiceProvider.php [hunk 3]
```diff
@@ -341,8 +342,12 @@ protected function parseConnections($options, $defaultHost, $defaultPort)
...
         } elseif (isset($options['save_path'])) {
             foreach (explode(',', $options['save_path']) as $conn) {
+                $uri = new Uri($conn);
+
                 $connBag = new ParameterBag();
-                $connBag->set('uri', new Uri($conn));
+                $connBag->set('uri', $uri);
+                $connBag->add(parse_query($uri->getQuery()));
```
Comment (dwolfhub) [inline src/Provider/SessionServiceProvider.php]: @CarsonF is this what you had in mind?

File: src/Session/Handler/RedisHandler.php [hunk 4]
```diff
@@ -25,7 +25,7 @@ class RedisHandler implements \SessionHandlerInterface, LazyWriteHandlerInterfac
      */
     public function __construct($redis, $maxlifetime)
     {
-        if (!$redis instanceof Redis || !$redis instanceof Predis) {
+        if (!$redis instanceof Redis && !$redis instanceof Predis) {
```
Comment (CarsonF) [inline src/Session/Handler/RedisHandler.php]: Who did that smh

Comment (GawainLynch) [inline src/Session/Handler/RedisHandler.php]: I know, right :wink:

**Comment by dwolfhub** [comment]
@CarsonF rebased and ready!

**Comment by CarsonF** [comment]
@GawainLynch Good for me :shipit:

**Comment by GawainLynch** [comment]
Merged into `release/3.1` via https://github.com/bolt/bolt/commit/dcd15150d07699e1f31b862bcd624619ba3ee7b8

**Pull request closed without merging**